src/StreamJsonRpc/Reflection/MessageFormatterProgressTracker.cs

```diff
@@ -0,0 +1,239 @@
+        /// </summary>
+        private readonly Dictionary<long, object> requestProgressMap = new Dictionary<long, object>();
+
+        /// <summary>
+        /// Dictionary used to map progress id token to its corresponding <see cref="ProgressParamInformation" /> instance containing the progress object and the necessary fields to report the results.
+        /// </summary>
+        private readonly Dictionary<object, ProgressParamInformation> progressMap = new Dictionary<object, ProgressParamInformation>();
+
+        /// <summary>
+        /// Object used to lock the access to <see cref="requestProgressMap"/> and <see cref="progressMap"/>.
+        /// </summary>
+        private readonly object progressLock = new object();
+
+        /// <summary>
+        /// Gets or Sets the id of the request currently being serialized so the converter can use it to create the request-progress map.
+        /// </summary>
+        public long? RequestIdBeingSerialized { get; set; }
+
+        /// <summary>
+        /// Gets or sets the the next id value to assign as token for the progress objects.
+        /// </summary>
+        private long NextProgressId { get; set; }
+
+        /// <summary>
+        /// Converts given <see cref="Type"/> to its <see cref="IProgress{T}"/> type.
+        /// </summary>
+        /// <param name="objectType">The type which may implement <see cref="IProgress{T}"/>.</param>
+        /// <returns>The <see cref="IProgress{T}"/> from given <see cref="Type"/> object, or <c>null</c>  if no such interface was found in the given <paramref name="objectType" />.</returns>
+        public static Type FindIProgressOfT(Type objectType)
+        {
+            Requires.NotNull(objectType, nameof(objectType));
+
+            if (objectType.IsConstructedGenericType && objectType.GetGenericTypeDefinition().Equals(typeof(IProgress<>)))
+            {
+                return objectType;
+            }
+
+            Type iProgressOfTType = null;
+            lock (IProgressOfTTypeMap)
+            {
+                if (!IProgressOfTTypeMap.TryGetValue(objectType, out iProgressOfTType))
+                {
+                    iProgressOfTType = objectType.GetTypeInfo().GetInterfaces().FirstOrDefault(i => i.IsConstructedGenericType && i.GetGenericTypeDefinition() == typeof(IProgress<>));
+                    IProgressOfTTypeMap.Add(objectType, iProgressOfTType);
+                }
+            }
+
+            return iProgressOfTType;
+        }
+
+        /// <summary>
+        /// Saves a <see cref="ProgressParamInformation"/> instance obtained from the given <see cref="object"/> into <see cref="progressMap"/>.
+        /// </summary>
+        /// <param name="value">The object which should implement <see cref="IProgress{T}"/> to create the <see cref="ProgressParamInformation"/> instance.</param>
+        /// <returns>The assigned <see cref="long"/> progres ID.</returns>
+        public long AddProgressObjectToMap(object value)
```

Comment:
>AddProgressObjectToMap [](start = 20, length = 22)

This method name describes more of the internals of what it does than what the outside caller should have to think about. Can you name this based on the use case for the caller? For example: `GetTokenForProgress`